pkg/vault/contracts/Authorizer.sol

```diff
@@ -63,16 +102,105 @@ contract Authorizer is IAuthorizer {
+    }
+
+    /**
+     * @dev Sets a new delay for `action`
+     */
+    function setDelay(bytes32 action, uint256 delay) external {
+        _require(msg.sender == address(this), Errors.SENDER_NOT_ALLOWED);
+        delays[action] = delay;
+        emit ActionDelaySet(action, delay);
+    }
+
+    /**
+     * @dev Schedules a delay change of `newDelay` for `action`
+     */
+    function scheduleDelayChange(
+        bytes32 action,
+        uint256 newDelay,
+        address[] memory executors
+    ) external returns (uint256 id) {
+        require(newDelay <= MAX_DELAY, "DELAY_TOO_LARGE");
+        bytes32 setDelayAction = keccak256(abi.encodePacked(SET_DELAY_PERMISSION, action));
+        _authenticate(setDelayAction, address(this));
+
+        uint256 actionDelay = delays[action];
+        bytes memory data = abi.encodeWithSelector(this.setDelay.selector, action, newDelay);
+        return _schedule(setDelayAction, address(this), data, actionDelay, executors);
+    }
+
+    /**
+     * @dev Schedules a new action
+     */
+    function schedule(
+        address where,
+        bytes memory data,
+        address[] memory executors
+    ) external returns (uint256 id) {
+        require(where != address(this), "CANNOT_SCHEDULE_AUTHORIZER_ACTIONS");
+        bytes32 action = IAuthentication(where).getActionId(_decodeSelector(data));
+        _require(hasPermission(action, msg.sender, where), Errors.SENDER_NOT_ALLOWED);
+
+        uint256 delay = delays[action];
+        require(delay > 0, "CANNOT_SCHEDULE_ACTION");
+        return _schedule(action, where, data, delay, executors);
+    }
+
+    /**
+     * @dev Executes action `id`
+     */
+    function execute(uint256 id) external returns (bytes memory result) {
+        require(id < scheduledActions.length, "ACTION_DOES_NOT_EXIST");
+        ScheduledAction storage scheduledAction = scheduledActions[id];
+        require(!scheduledAction.executed, "ACTION_ALREADY_EXECUTED");
+        require(!scheduledAction.cancelled, "ACTION_ALREADY_CANCELLED");
+
+        // solhint-disable-next-line not-rely-on-time
+        require(block.timestamp >= scheduledAction.executableAt, "ACTION_NOT_EXECUTABLE");
+        if (scheduledAction.protected) {
+            _authenticate(_executeActionId(id), address(this));
+        }
+
+        scheduledAction.executed = true;
+        result = scheduledAction.where.functionCall(scheduledAction.data);
+        emit ActionExecuted(id);
+    }
+
+    /**
+     * @dev Cancels action `id`
+     */
+    function cancel(uint256 id) external {
+        require(id < scheduledActions.length, "ACTION_DOES_NOT_EXIST");
+        ScheduledAction storage scheduledAction = scheduledActions[id];
+
+        require(!scheduledAction.executed, "ACTION_ALREADY_EXECUTED");
+        require(!scheduledAction.cancelled, "ACTION_ALREADY_CANCELLED");
+        _require(msg.sender == scheduledAction.sender, Errors.SENDER_NOT_ALLOWED);
```

Comment:
I didn't want to introduce another permission, maybe we can work with the "executors" as well? Instead of keeping the creator only, use the same logic we do for the executors

Reply:
Hm no, I don't think executors being able to cancel would be a good fit. And we definitely don't want public cancellation if there are no executors.

Using the schedule permission for cancellation wouldn't result in new permissions, would it? I'm talking about having the same logic as the one we have in `schedule`:

```
        bytes32 action = IAuthentication(where).getActionId(_decodeSelector(data));
        _require(hasPermission(action, msg.sender, where), Errors.SENDER_NOT_ALLOWED);
```

Since we can grant and revoke that permission, we have flexibility over cancellations. On the other hand, if it is a contract that schedules an action, or if the keys of the creator are somehow lost, it'd be impossible to cancel.